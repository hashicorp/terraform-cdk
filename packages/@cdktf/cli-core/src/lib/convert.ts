/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

import * as hcl2cdk from "@cdktf/hcl2cdk";
import { CdktfConfig } from "./cdktf-config";
import {
  ConstructsMakerProviderTarget,
  TerraformProviderConstraint,
  readSchema,
} from "@cdktf/provider-generator";
import { Errors, LANGUAGES } from "@cdktf/commons";

export async function convertConfigurationFile(configuration: string) {
  const cfg = CdktfConfig.read(process.cwd()); // TODO: make this the project directory instead of cwd
  const targets = cfg.terraformProviders.map((constraint) =>
    ConstructsMakerProviderTarget.from(
      new TerraformProviderConstraint(constraint),
      LANGUAGES[0]
    )
  );
  const { providerSchema } = await readSchema(targets);
  if (!providerSchema) {
    throw Errors.Internal("Could not fetch provider schema");
  }

  const { all } = await hcl2cdk.convert(configuration, {
    providerSchema,
    language: cfg.language,
    codeContainer: "constructs.Construct",
  });

  return all;
}
