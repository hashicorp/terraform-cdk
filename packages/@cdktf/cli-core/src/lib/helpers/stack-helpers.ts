// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0

import { ensureAllSettledBeforeThrowing, Errors, logger } from "@cdktf/commons";
import minimatch from "minimatch";
import { CdktfStack } from "../cdktf-stack";
import { SynthesizedStack } from "../synth-stack";

export function getSingleStack(
  stacks: SynthesizedStack[],
  stackName?: string,
  targetAction?: string,
) {
  if (!stacks) {
    throw Errors.Internal(
      "Trying to access a stack before it has been synthesized",
    );
  }

  if (stackName) {
    const stack = stacks.find((s) => s.name === stackName);
    if (!stack) {
      throw Errors.Usage("Could not find stack: " + stackName);
    }

    return stack;
  }

  if (stacks.length === 1) {
    return stacks[0];
  }

  throw Errors.Usage(
    `Found more than one stack, please specify a target stack. Run cdktf ${
      targetAction || "<verb>"
    } <stack> with one of these stacks: ${stacks
      .map((s) => s.name)
      .join(", ")} `,
  );
}

export function getMultipleStacks(
  stacks: SynthesizedStack[],
  patterns?: string[],
  targetAction?: string,
) {
  if (!patterns || !patterns.length) {
    if (stacks.length === 1) {
      return [stacks[0]];
    }
    throw Errors.Usage(
      `Found more than one stack, please specify a target stack. Run cdktf ${
        targetAction || "<verb>"
      } <stack> with one of these stacks: ${stacks
        .map((s) => s.name)
        .join(", ")} `,
    );
  }

  return patterns.flatMap((pattern) => {
    const matchingStacks = stacks.filter((stack) =>
      minimatch(stack.name, pattern),
    );

    if (matchingStacks.length === 0) {
      throw Errors.Usage(`Could not find stack for pattern '${pattern}'`);
    }

    return matchingStacks;
  });
}

// Returns the first stack that has no unmet dependencies
// An unmet dependency is a dependency that has not been deployed yet
// If there is no unfinished stack, returns undefined
// If there is no stack ready to be worked on, it returns a promise that will resolve as soon as there is a follow-up stack available
export async function getStackWithNoUnmetDependencies(
  stackExecutors: CdktfStack[],
): Promise<CdktfStack | undefined> {
  logger.debug("Checking for stacks with unmet dependencies");
  logger.debug("stack executors:", stackExecutors);
  const pendingStacks = stackExecutors.filter((executor) => executor.isPending);
  logger.debug("pending stacks:", stackExecutors);

  if (pendingStacks.length === 0) {
    return undefined;
  }

  const currentlyReadyStack = pendingStacks.find((executor) =>
    executor.stack.dependencies.every(
      (dependency) =>
        stackExecutors.find((executor) => executor.stack.name === dependency)
          ?.currentState === "done",
    ),
  );

  if (currentlyReadyStack) {
    logger.debug("Found a stack with no unmet dependencies");
    return currentlyReadyStack;
  }

  const stackExecutionPromises = stackExecutors
    .filter((ex) => ex.currentWorkPromise)
    .map((ex) => ex.currentWorkPromise);

  logger.debug(
    `${stackExecutionPromises.length} stacks are currently busy, waiting for one to finish`,
  );

  if (!stackExecutionPromises.length) {
    return undefined;
  }

  await ensureAllSettledBeforeThrowing(
    Promise.race(stackExecutionPromises),
    stackExecutionPromises,
  );

  return await getStackWithNoUnmetDependencies(stackExecutors);
}

function findAllDependantStacks(
  stackExecutors: CdktfStack[],
  stackName: string,
): CdktfStack[] {
  return stackExecutors.filter((innerExecutor) =>
    innerExecutor.stack.dependencies.includes(stackName),
  );
}

export function findAllNestedDependantStacks(
  stackExecutors: CdktfStack[],
  stackName: string,
  knownDependantStackNames: Set<string> = new Set(),
): CdktfStack[] {
  const dependantStacks = findAllDependantStacks(stackExecutors, stackName);
  dependantStacks.forEach((stack) => {
    if (knownDependantStackNames.has(stack.stack.name)) {
      return;
    }

    knownDependantStackNames.add(stack.stack.name);
    findAllNestedDependantStacks(
      stackExecutors,
      stack.stack.name,
      knownDependantStackNames,
    );
  });

  return stackExecutors.filter((executor) =>
    knownDependantStackNames.has(executor.stack.name),
  );
}

// Returns the first stack that has no dependents that need to be destroyed first
export async function getStackWithNoUnmetDependants(
  stackExecutors: CdktfStack[],
): Promise<CdktfStack | undefined> {
  logger.debug("Checking for stacks with unmet dependants");
  logger.debug("stack executors:", stackExecutors);
  const pendingStacks = stackExecutors.filter((executor) => executor.isPending);
  logger.debug("pending stacks:", stackExecutors);

  if (pendingStacks.length === 0) {
    return undefined;
  }

  const currentlyReadyStack = pendingStacks.find((executor) => {
    const dependantStacks = findAllDependantStacks(
      stackExecutors,
      executor.stack.name,
    );
    return dependantStacks.every((stack) => stack.currentState === "done");
  });

  if (currentlyReadyStack) {
    logger.debug("Found a stack with no unmet dependants");
    return currentlyReadyStack;
  }
  const stackExecutionPromises = stackExecutors
    .filter((ex) => ex.currentWorkPromise)
    .map((ex) => ex.currentWorkPromise);

  logger.debug(
    `${stackExecutionPromises.length} stacks are currently busy, waiting for one to finish`,
  );
  if (!stackExecutionPromises.length) {
    return undefined;
  }

  await Promise.race(stackExecutionPromises);
  return await getStackWithNoUnmetDependants(stackExecutors);
}

// Throws an error if there is a dependant stack that is not included
export function checkIfAllDependantsAreIncluded(
  stacksToRun: SynthesizedStack[],
  allStacks: SynthesizedStack[],
) {
  const allDependants = new Set<string>();
  stacksToRun
    .map((stack) =>
      allStacks.filter((s) => s.dependencies.includes(stack.name)),
    )
    .flat()
    .forEach((dependant) => allDependants.add(dependant.name));

  const stackNames = stacksToRun.map((stack) => stack.name);
  const missingDependants = [...allDependants].filter(
    (dependant) => !stackNames.includes(dependant),
  );

  if (missingDependants.length > 0) {
    throw Errors.Usage(
      `The following dependant stacks are not included in the stacks to run: ${missingDependants.join(
        ", ",
      )}. Either add them or add the --ignore-missing-stack-dependencies flag.`,
    );
  }
}

/** 
  Throws an error if there is a dependency that is not included
  Cycles are detected on dependency creation at synthesis time
  Running this prevents us from being in a situation where we 
  have to wait for a stack to be deployed that is not included to be run
*/
export function checkIfAllDependenciesAreIncluded(
  stacksToRun: SynthesizedStack[],
) {
  const allDependencies = new Set<string>();
  stacksToRun
    .map((stack) => stack.dependencies)
    .flat()
    .forEach((dependency) => allDependencies.add(dependency));

  const stackNames = stacksToRun.map((stack) => stack.name);
  const missingDependencies = [...allDependencies].filter(
    (dependency) => !stackNames.includes(dependency),
  );

  if (missingDependencies.length > 0) {
    throw Errors.Usage(
      `The following dependencies are not included in the stacks to run: ${missingDependencies.join(
        ", ",
      )}. Either add them or add the --ignore-missing-stack-dependencies flag.`,
    );
  }
}
