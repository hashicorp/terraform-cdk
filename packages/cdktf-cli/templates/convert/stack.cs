using System;
using Constructs;
using HashiCorp.Cdktf;
/* {{ imports }} */

namespace MyCompany.MyApp
{
    class MyStack : TerraformStack
    {
        public MyStack(Construct scope, string id) : base(scope, id)
        {
            /* {{ code }} */
        }
    }
}