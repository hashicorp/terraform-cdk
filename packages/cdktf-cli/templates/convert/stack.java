package com.mycompany.app;

import software.constructs.Construct;
import com.hashicorp.cdktf.TerraformStack;
/* {{ imports }} */

public class MyStack extends TerraformStack
{
    public MyStack(final Construct scope, final String id) {
        super(scope, id);

        /* {{ code }} */
    }
}