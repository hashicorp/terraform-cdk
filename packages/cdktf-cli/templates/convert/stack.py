#!/usr/bin/env python
# Copyright (c) HashiCorp, Inc.
# SPDX-License-Identifier: MPL-2.0

from constructs import Construct
from cdktf import TerraformStack
# {{ imports }}


class MyStack(TerraformStack):
    def __init__(self, scope: Construct, id: str):
        super().__init__(scope, id)

        # {{ code }}
