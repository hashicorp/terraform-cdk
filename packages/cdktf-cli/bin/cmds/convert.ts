import yargs from "yargs";
import { Errors } from "../../lib/errors";
import { requireHandlers } from "./helper/utilities";

class Command implements yargs.CommandModule {
  public readonly command = "convert [OPTIONS]";
  public readonly describe =
    "Converts a single file of HCL configuration to CDK for Terraform. Takes the file to be converted on stdin.";

  public readonly builder = (args: yargs.Argv) =>
    args
      .example(
        "cat main.tf | cdktf convert  --provider integrations/github",
        "Takes the HCL content of main.tf and converts it to CDK for Terraform content and prints it"
      )
      .example(
        "cat main.tf | cdktf convert --provider hashicorp/aws > imported.ts",
        "Takes the HCL content of main.tf and converts it to CDK for Terraform content in imported.ts"
      )
      .example(
        "cat main.tf | cdktf convert --provider 'hashicorp/aws@ ~>3.62.0' 'integrations/github@ ~>4.16.0' --language python > imported.py",
        "Takes the HCL content of main.tf and converts it to CDK for Terraform content in imported.ts"
      )
      .option("language", {
        choices: ["typescript", "python", "csharp", "java"],
        default: "typescript",
      })
      .option("provider", {
        describe:
          "The conversion needs to know which providers are used in addition to the ones in your cdktf.json file. We search for a cdktf.json below your current working directory.",
        type: "array",
        default: [],
      })
      .showHelpOnFail(true);

  public async handler(argv: any) {
    Errors.setScope("convert");
    // deferred require to keep cdktf-cli main entrypoint small (e.g. for fast shell completions)
    const api = requireHandlers();

    await api.convert(argv);
  }
}

module.exports = new Command();
