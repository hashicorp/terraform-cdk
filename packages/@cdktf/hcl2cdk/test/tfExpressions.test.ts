/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

import { testCase, Synth, binding, Snapshot } from "./helpers/convert";

describe("tfExpressions", () => {
  testCase.test(
    "arithmetics",
    `
          variable "members" {
            type = number
          }
          variable "admins" {
            type = number
          }

          output "arithmetics" {
            value = var.members + var.admins
          }`,
    [],
    Snapshot.yes,
    Synth.yes
  );

  testCase.test(
    "conditionals",
    `
      provider "aws" {
        region                      = "us-east-1"
      }
      resource "aws_kms_key" "examplekms" {
        description             = "KMS key 1"
        deletion_window_in_days = 7
      }

      resource "aws_s3_bucket" "examplebucket" {
        bucket = "examplebuckettftest"
        acl    = "private"
      }

      resource "aws_s3_bucket_object" "examplebucket_object" {
        key        = "someobject"
        bucket     = aws_kms_key.examplekms.deletion_window_in_days > 3 ? aws_s3_bucket.examplebucket.id : []
        source     = "index.html"
        kms_key_id = aws_kms_key.examplekms.arn
      }`,
    [binding.aws],
    Snapshot.yes,
    Synth.yes,
    {
      resources: ["aws_s3_bucket", "aws_kms_key", "aws_s3_bucket_object"],
    }
  );

  testCase.test(
    "for expression 1",
    `
            variable "users" {
              type = map(object({
                is_admin = boolean
              }))
            }

            locals {
              admin_users = {
                for name, user in var.users : name => user
                if user.is_admin
              }
              regular_users = {
                for name, user in var.users : name => user
                if !user.is_admin
              }
            }

            output "combined-so-it-does-not-get-removed" {
              value = "\${local.admin_users},\${local.regular_users}"
            }
            `,
    [],
    Snapshot.yes,
    Synth.yes
  );

  testCase.test(
    "for expression 2",
    `
          variable "users" {
            type = map(object({
              role = string
            }))
          }

          locals {
            users_by_role = {
              for name, user in var.users : user.role => name...
            }
          }

          output "so-it-does-not-get-removed" {
            value = local.users_by_role
          }`,
    [],
    Snapshot.yes,
    Synth.yes
  );

  testCase.test(
    "for expression 3",
    `
            provider "datadog" {
              api_key = "api_key"
              app_key = "app_key"
            }

            variable "users" {
              type = list(object({
                id = string
              }))
            }

            resource "datadog_monitor" "hard_query" {
              name    = "queries are hard"
              message = "here we go"
              query   = join(" && ", [for o in var.users : "!(!\${o.id})"])
              type    = "metric alert"
            }`,
    [binding.datadog],
    Snapshot.yes,
    Synth.yes,
    {
      resources: ["datadog_monitor"],
    }
  );

  testCase.test(
    "property access through square brackets",
    `
      provider "aws" {
        region                      = "us-east-1"
      }
      variable "settings" {
        type = map(string)
      }

      resource "aws_s3_bucket" "examplebucket" {
        bucket = var.settings["bucket_name"]
        acl    = "private"
      }
      `,
    [binding.aws],
    Snapshot.yes,
    Synth.yes,
    {
      resources: ["aws_s3_bucket"],
    }
  );

  testCase.test(
    "list access through square brackets",
    `
      provider "aws" {
        region                      = "us-east-1"
      }
      variable "settings" {
        type = list(map(string))
      }

      resource "aws_s3_bucket" "examplebucket" {
        bucket = var.settings[0]["bucket_name"]
        acl    = "private"
      }
      `,
    [binding.aws],
    Snapshot.yes,
    Synth.yes,
    {
      resources: ["aws_s3_bucket"],
    }
  );

  testCase.test(
    "numeric property access",
    `
      provider "google" {
        project = "my-project"
        region  = "us-central1"
      }
      resource "google_compute_instance" "example" {
        name          = "example"
        machine_type  = "f1-micro"
        zone          = "us-east1-b"

        boot_disk {
          initialize_params {
            image = "debian-cloud/debian-9"
          }
        }

        network_interface {
          network = "default"

          access_config {
            // Ephemeral IP
          }
        }
      }

      output "public_ip" {
        value = "\${google_compute_instance.example.network_interface.0.access_config.0.assigned_nat_ip}"
      }
      `,
    [binding.google],
    Snapshot.yes,
    Synth.yes,
    {
      resources: ["google_compute_instance"],
    }
  );

  testCase.test(
    "multi-line strings are supported",
    `
      output "hash" {
        value = <<ITEM
{
  "example_hash_key": {"S": "something"},
  "example_attribute": {"N": "11111"}
}
ITEM
      }
      `,
    [],
    Snapshot.yes,
    Synth.yes,
    {
      resources: [],
    }
  );

  testCase.test(
    "escaping dollar-curly in template strings works",
    `
    resource "aws_ssoadmin_instance_access_control_attributes" "example" {
      attribute {
        value {
          source = ["$\${path:name.givenName}"]
        }
      }
    }
      `,
    [binding.aws],
    Snapshot.yes,
    Synth.yes_but_only_typescript_right_now_because_it_breaks,
    {
      resources: [],
    }
  );

  testCase.test(
    "property access in maps works",
    `
    variable "default_tags" {
      type        = map(string)
      description = "Map of default tags to apply to resources"
      default = {
        project = "Learning Live with AWS & HashiCorp"
      }
    }
    
    resource "aws_eip" "nat" {
      vpc = true
      tags = {
        "Name" = "\${var.default_tags.project}-nat-eip"
      }
    }
      `,
    [binding.aws],
    Snapshot.yes_all_languages,
    Synth.yes_but_only_typescript_right_now_because_it_breaks,
    {
      resources: [],
    }
  );

  testCase.test(
    "strings containing single outer quotes are supported",
    `
      output "hash" {
        value = "'static'"
      }`,
    [],
    Snapshot.yes,
    Synth.yes,
    {
      resources: [],
    }
  );
});
