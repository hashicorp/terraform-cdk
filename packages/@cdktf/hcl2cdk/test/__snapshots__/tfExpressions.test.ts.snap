// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tfExpressions arithmetics snapshot typescript 1`] = `
"import { Construct } from "constructs";
import {
  VariableType,
  TerraformVariable,
  TerraformOutput,
  Op,
  TerraformStack,
} from "cdktf";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    const admins = new TerraformVariable(this, "admins", {
      type: VariableType.NUMBER,
    });
    const members = new TerraformVariable(this, "members", {
      type: VariableType.NUMBER,
    });
    new TerraformOutput(this, "arithmetics", {
      value: Op.add(members.value, admins.value),
    });
  }
}
"
`;

exports[`tfExpressions conditionals snapshot typescript 1`] = `
"import { Construct } from "constructs";
import { Op, conditional, Token, TerraformStack } from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { KmsKey } from "./.gen/providers/aws/kms-key";
import { AwsProvider } from "./.gen/providers/aws/provider";
import { S3Bucket } from "./.gen/providers/aws/s3-bucket";
import { S3BucketObject } from "./.gen/providers/aws/s3-bucket-object";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    new AwsProvider(this, "aws", {
      region: "us-east-1",
    });
    const examplekms = new KmsKey(this, "examplekms", {
      deletionWindowInDays: 7,
      description: "KMS key 1",
    });
    const examplebucket = new S3Bucket(this, "examplebucket", {
      acl: "private",
      bucket: "examplebuckettftest",
    });
    new S3BucketObject(this, "examplebucket_object", {
      bucket: Token.asString(
        conditional(
          Op.gt(examplekms.deletionWindowInDays, 3),
          examplebucket.id,
          []
        )
      ),
      key: "someobject",
      kmsKeyId: examplekms.arn,
      source: "index.html",
    });
  }
}
"
`;

exports[`tfExpressions escaping dollar-curly in template strings works snapshot typescript 1`] = `
"import { Construct } from "constructs";
import { TerraformStack } from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { SsoadminInstanceAccessControlAttributes } from "./.gen/providers/aws/ssoadmin-instance-access-control-attributes";
interface MyConfig {
  key: any;
  instanceArn: any;
}
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string, config: MyConfig) {
    super(scope, name);
    new SsoadminInstanceAccessControlAttributes(this, "example", {
      attribute: [
        {
          value: [
            {
              source: ["$\${path:name.givenName}"],
            },
          ],
          key: config.key,
        },
      ],
      instanceArn: config.instanceArn,
    });
  }
}
"
`;

exports[`tfExpressions for expression 1 snapshot typescript 1`] = `
"import { Construct } from "constructs";
import {
  VariableType,
  TerraformVariable,
  TerraformOutput,
  TerraformStack,
} from "cdktf";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    const users = new TerraformVariable(this, "users", {
      type: VariableType.map(
        VariableType.object({
          is_admin: VariableType.ANY,
        })
      ),
    });
    const adminUsers =
      "\${{ for name, user in \${" +
      users.value +
      "} : name => user if user.is_admin}}";
    const regularUsers =
      "\${{ for name, user in \${" +
      users.value +
      "} : name => user if !user.is_admin}}";
    new TerraformOutput(this, "combined-so-it-does-not-get-removed", {
      value: "\${" + adminUsers + "},\${" + regularUsers + "}",
    });
  }
}
"
`;

exports[`tfExpressions for expression 2 snapshot typescript 1`] = `
"import { Construct } from "constructs";
import {
  VariableType,
  TerraformVariable,
  TerraformOutput,
  TerraformStack,
} from "cdktf";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    const users = new TerraformVariable(this, "users", {
      type: VariableType.map(
        VariableType.object({
          role: VariableType.STRING,
        })
      ),
    });
    const usersByRole =
      "\${{ for name, user in \${" + users.value + "} : user.role => name...}}";
    new TerraformOutput(this, "so-it-does-not-get-removed", {
      value: usersByRole,
    });
  }
}
"
`;

exports[`tfExpressions for expression 3 snapshot typescript 1`] = `
"import { Construct } from "constructs";
import {
  VariableType,
  TerraformVariable,
  Fn,
  Token,
  TerraformStack,
} from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { Monitor } from "./.gen/providers/datadog/monitor";
import { DatadogProvider } from "./.gen/providers/datadog/provider";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    new DatadogProvider(this, "datadog", {
      apiKey: "api_key",
      appKey: "app_key",
    });
    const users = new TerraformVariable(this, "users", {
      type: VariableType.list(
        VariableType.object({
          id: VariableType.STRING,
        })
      ),
    });
    new Monitor(this, "hard_query", {
      message: "here we go",
      name: "queries are hard",
      query: Token.asString(
        Fn.join(
          " && ",
          Token.asList("\${[ for o in \${" + users.value + '} : "!(!\${o.id})"]}')
        )
      ),
      type: "metric alert",
    });
  }
}
"
`;

exports[`tfExpressions list access through square brackets snapshot typescript 1`] = `
"import { Construct } from "constructs";
import {
  VariableType,
  TerraformVariable,
  Fn,
  Token,
  TerraformStack,
} from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { AwsProvider } from "./.gen/providers/aws/provider";
import { S3Bucket } from "./.gen/providers/aws/s3-bucket";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    new AwsProvider(this, "aws", {
      region: "us-east-1",
    });
    const settings = new TerraformVariable(this, "settings", {
      type: VariableType.list(VariableType.map(VariableType.STRING)),
    });
    new S3Bucket(this, "examplebucket", {
      acl: "private",
      bucket: Token.asString(
        Fn.lookupNested(settings.value, ["0", '"bucket_name"'])
      ),
    });
  }
}
"
`;

exports[`tfExpressions multi-line strings are supported snapshot typescript 1`] = `
"import { Construct } from "constructs";
import { TerraformOutput, TerraformStack } from "cdktf";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    new TerraformOutput(this, "hash", {
      value:
        '{\\n  "example_hash_key": {"S": "something"},\\n  "example_attribute": {"N": "11111"}\\n}\\n\\n',
    });
  }
}
"
`;

exports[`tfExpressions numeric property access snapshot typescript 1`] = `
"import { Construct } from "constructs";
import { TerraformOutput, Fn, TerraformStack } from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { ComputeInstance } from "./.gen/providers/google/compute-instance";
import { GoogleProvider } from "./.gen/providers/google/provider";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    new GoogleProvider(this, "google", {
      project: "my-project",
      region: "us-central1",
    });
    const example = new ComputeInstance(this, "example", {
      bootDisk: {
        initializeParams: {
          image: "debian-cloud/debian-9",
        },
      },
      machineType: "f1-micro",
      name: "example",
      networkInterface: [
        {
          accessConfig: [{}],
          network: "default",
        },
      ],
      zone: "us-east1-b",
    });
    new TerraformOutput(this, "public_ip", {
      value: Fn.lookupNested(example.networkInterface, [
        "0",
        "access_config",
        "0",
        "assigned_nat_ip",
      ]),
    });
  }
}
"
`;

exports[`tfExpressions property access in maps works snapshot csharp 1`] = `
"using Constructs;
using HashiCorp.Cdktf;
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
using aws.Eip;
class MyConvertedCode : TerraformStack
{
    public MyConvertedCode(Construct scope, string name) : base(scope, name)
    {
        /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
            You can read more about this at https://cdk.tf/variables*/
        var defaultTags = new TerraformVariable(this, "default_tags", new TerraformVariableConfig {
            Default = new [] { new Dictionary<string, string> {
                { "project", "Learning Live with AWS & HashiCorp" }
            } },
            Description = "Map of default tags to apply to resources",
            Type = VariableType.Map(VariableType.STRING)
        });
        new Eip(this, "nat", new EipConfig {
            Tags = new Dictionary<string, string> {
                { "Name", Token.AsString(Fn.LookupNested(defaultTags.Value, new [] { "project" })) + "-nat-eip" }
            },
            Vpc = true
        });
    }
}"
`;

exports[`tfExpressions property access in maps works snapshot go 1`] = `
"import "github.com/aws/constructs-go/constructs"
import "github.com/hashicorp/terraform-cdk-go/cdktf"
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import "cdk.tf/go/stack/generated/aws/eip"
type myConvertedCode struct {
	terraformStack
}

func newMyConvertedCode(scope construct, name *string) *myConvertedCode {
	this := &myConvertedCode{}
	newTerraformStack_Override(this, scope, name)
	/*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
	    You can read more about this at https://cdk.tf/variables*/
	defaultTags := cdktf.NewTerraformVariable(this, jsii.String("default_tags"), &TerraformVariableConfig{
		Default: []interface{}{
			map[string]*string{
				"project": jsii.String("Learning Live with AWS & HashiCorp"),
			},
		},
		Description: jsii.String("Map of default tags to apply to resources"),
		Type: cdktf.VariableType_Map(cdktf.VariableType_STRING()),
	})
	genprovidersawseip.NewEip(this, jsii.String("nat"), &eipConfig{
		tags: map[string]*string{
			"Name": jsii.String(cdktf.Token_asString(cdktf.Fn_lookupNested(defaultTags.value, []interface{}{
				jsii.String("project"),
			})) + "-nat-eip"),
		},
		vpc: jsii.Boolean(true),
	})
	return this
}"
`;

exports[`tfExpressions property access in maps works snapshot java 1`] = `
"import software.constructs.Construct;
import com.hashicorp.cdktf.VariableType;
import com.hashicorp.cdktf.TerraformVariable;
import com.hashicorp.cdktf.Fn;
import com.hashicorp.cdktf.Token;
import com.hashicorp.cdktf.TerraformStack;
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import imports.aws.eip.Eip;
public class MyConvertedCode extends TerraformStack {
    public MyConvertedCode(Construct scope, String name) {
        super(scope, name);
        /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
            You can read more about this at https://cdk.tf/variables*/
        TerraformVariable defaultTags = TerraformVariable.Builder.create(this, "default_tags")
                .default(List.of(Map.of(
                        "project", "Learning Live with AWS & HashiCorp")))
                .description("Map of default tags to apply to resources")
                .type(VariableType.map(VariableType.STRING))
                .build();
        new Eip(this, "nat", new EipConfig()
                .tags(Map.of(
                        "Name", Token.asString(Fn.lookupNested(defaultTags.getValue(), List.of("project"))) + "-nat-eip"))
                .vpc(true)
                );
    }
}"
`;

exports[`tfExpressions property access in maps works snapshot python 1`] = `
"from constructs import Construct
from cdktf import VariableType, TerraformVariable, Fn, Token, TerraformStack
#
# Provider bindings are generated by running \`cdktf get\`.
# See https://cdk.tf/provider-generation for more details.
#
from imports.aws.eip import Eip
class MyConvertedCode(TerraformStack):
    def __init__(self, scope, name):
        super().__init__(scope, name)
        # Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
        #     You can read more about this at https://cdk.tf/variables
        default_tags = TerraformVariable(self, "default_tags",
            default=[{
                "project": "Learning Live with AWS & HashiCorp"
            }
            ],
            description="Map of default tags to apply to resources",
            type=VariableType.map(VariableType.STRING)
        )
        Eip(self, "nat",
            tags={
                "Name":
                Token.as_string(Fn.lookup_nested(default_tags.value, ["project"])) + "-nat-eip"
            },
            vpc=True
        )"
`;

exports[`tfExpressions property access in maps works snapshot typescript 1`] = `
"import { Construct } from "constructs";
import {
  VariableType,
  TerraformVariable,
  Fn,
  Token,
  TerraformStack,
} from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { Eip } from "./.gen/providers/aws/eip";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    const defaultTags = new TerraformVariable(this, "default_tags", {
      default: [
        {
          project: "Learning Live with AWS & HashiCorp",
        },
      ],
      description: "Map of default tags to apply to resources",
      type: VariableType.map(VariableType.STRING),
    });
    new Eip(this, "nat", {
      tags: {
        Name:
          Token.asString(Fn.lookupNested(defaultTags.value, ["project"])) +
          "-nat-eip",
      },
      vpc: true,
    });
  }
}
"
`;

exports[`tfExpressions property access through square brackets snapshot typescript 1`] = `
"import { Construct } from "constructs";
import {
  VariableType,
  TerraformVariable,
  Fn,
  Token,
  TerraformStack,
} from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { AwsProvider } from "./.gen/providers/aws/provider";
import { S3Bucket } from "./.gen/providers/aws/s3-bucket";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    new AwsProvider(this, "aws", {
      region: "us-east-1",
    });
    const settings = new TerraformVariable(this, "settings", {
      type: VariableType.map(VariableType.STRING),
    });
    new S3Bucket(this, "examplebucket", {
      acl: "private",
      bucket: Token.asString(
        Fn.lookupNested(settings.value, ['"bucket_name"'])
      ),
    });
  }
}
"
`;

exports[`tfExpressions strings containing single outer quotes are supported snapshot typescript 1`] = `
"import { Construct } from "constructs";
import { TerraformOutput, TerraformStack } from "cdktf";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    new TerraformOutput(this, "hash", {
      value: "'static'",
    });
  }
}
"
`;
